tests/suites/os/tests/secureboot/index.js

```diff
@@ -1,145 +1,339 @@
-							'"apk add --update --no-cache binutils && strip /mnt/boot/EFI/BOOT/bootx64.efi"'],
-						this.link,
-					).then(() => this.worker.executeCommandInHostOS('reboot', this.link)
-					).then(() => test.resolves(
-							waitForSerialOutput(/bootx64.efi: Access Denied/),
-							'Firmware will not load bootloader that fails verification',
-						)
-					).then(resetWorker);
-				};
-
-				const testBootloaderConfigIntegrity = async () => {
-					return this.worker.executeCommandInHostOS(
-						['sed', '-i', 's/lockdown=integrity//', '/mnt/efi/EFI/BOOT/grub.cfg'],
-						this.link
-					).then(() => this.worker.executeCommandInHostOS('reboot', this.link)
-					).then(() => test.resolves(
-						/* The below pattern is the expected output when the config file
-						 * fails the signature check and the fallback console is not enabled.
-						 *
-						 * This sequence of ANSI escape codes decodes as:
-						 *
-						 *	ESC[0m - reset all modes
-						 *	ESC[37m - white foreground
-						 *	ESC[40m - black background
-						 *
-						 *	https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797
-						 *
-						 * Otherwise, the output finishes the above background color escape
-						 * sequence, clears the screen, and shows 'loading Boot0003
-						 * "balenaOS"'
-						 */
-
-							waitForSerialOutput(/\x1B\[0m\x1B\[37m\x1B\[40m$/, 3), // eslint-disable-line no-control-regex
-							'Bootloader will not load configuration that fails signature verification',
-						{ todo: 'This needs reworked for GRUB 2.12' },
-						)
-					).then(resetWorker);
-				};
-
-				if (securebootSupported) {
-					if (securebootEnabled) {
-						await testFullDiskEncryption();
-						await testModuleVerification();
-						await testBootloaderIntegrity();
-						await testBootloaderConfigIntegrity();
-					} else {
-						test.comment('Secure boot is not enabled');
-					}
-				} else {
-					test.comment('Secure boot is not supported by firmware');
+	async testModuleVerification() {
+		if (this.module.name.length != 0) {
+			await this.test.resolves(
+				this.worker.executeCommandInHostOS(`modprobe ${this.module.name}`, this.link),
+				`Module ${this.module.name} with valid signature loads`,
+			);
+		} else {
+			this.test.comment("No module with valid signature specified - skipping")
+		}
+
+		if (this.module.headersVersion.length != 0) {
+			const srcDir = `${__dirname}/kernel-module-build/`
+			await fse.copy(srcDir, this.tmpDir);
+			fs.readFile(`${this.tmpDir}/docker-compose.yml`, 'utf-8', (err,data) => {
+				if (err) {
+					throw new Error(`Unable to find ${this.tmpDir}/docker-compose.yml`)
 				}
+				const result = data.replace(/os_version:\s*\S+/, `os_version: ${this.module.headersVersion}`);
+				fs.writeFile( `${this.tmpDir}/docker-compose.yaml`, result, 'utf-8', (err) => {
+					if (err) {
+						throw new Error(`Unable to write ${this.tmpDir}/docker-compose.yml`)
+					}
+				})
+			})
```

Comment:
Would make the same suggestion here as Ryan did one before this, writing this file at this point synchronously would be best.

Additional suggestion in both Ryan's comment and this would be error handling. CHeckout this psuedo code for this.

```suggestion
     		try {
     		  await fse.writeFile(`${this.tmpDir}/docker-compose.yaml`, result, 'utf-8')
     		 } catch (err) {
						throw new Error(`Unable to write ${this.tmpDir}/docker-compose.yml: ${err}`)
			 }
```

Reply:
This has gone outdated, as I see you just pushed your commit @alexgg as I was reviewing. If you want you can implement the error handling like this to catch the errors and gain back the feature parity that you get with callbacks.

Reply:
@vipulgupta2048 I thought about it, but without the try/catch the library exception will surface instead which will have more details that the thrown error exception. Is it best practise not to raise the original exception?

Reply:
Indeed, the exception will surface with a traceback telling you exactly what happened. I went into the rabbit hole to find the best practice 

The basic rule of thumb for catching exceptions is to catch exceptions if and only if you have a meaningful way of handling them.

> Don't catch an exception if you're only going to log the exception and throw it up the stack. It serves no meaning and clutters code. Do catch an exception when you are expecting a failure in a specific part of your code, and if you have a fallback for it.